ci(aio): verify that the Lighthouse PWA score remains above a threshold (and other fixes)

gkalpak opened the pull request:
1. Better message for disabled JavaScript.
2. Allow reloading page from ServiceWorker update notification. This helps when running in "standalone" mode on mobile as a PWA, where it is not obvious or possible to refresh by other means.
3. Verify that the [Lighthouse][1] PWA score remains above a threshold (currently 90) on CI.

[1]: https://developers.google.com/web/tools/lighthouse/

gkalpak committed:
fix(aio): add better message for disabled JavaScript

gkalpak committed:
feat(aio): allow reloading page from ServiceWorker update notification

gkalpak committed:
ci(aio): verify that the Lighthouse PWA score remains above a threshold

gkalpak commented on aio/src/index.html:
```diff
@@ -33,6 +33,19 @@
+        <img src="assets/images/logos/angular2/angular.svg" />
+        <h1>One framework.<br>Mobile &amp; desktop.</h1>
+        <br />
+        <h3 style="color:red">
```
This might need styling. Left it for @sjtrimble :smiley:

gkalpak commented on aio/src/app/sw-updates/global.value.ts:
```diff
@@ -0,0 +1,5 @@
+
+
+export const Global = new InjectionToken<Window>('global');
+export const globalProvider = { provide: Global, useValue: window };
```
Had to use a provider so that it can be mocked in tests. `window.location.reload` in non-writable, so I couldn't spy on it or temporarily overwrite.

gkalpak commented on aio/scripts/test-pwa-score.js:
```diff
@@ -0,0 +1,62 @@
+  // Since the tests are run on a local server,
+  // ignore HTTPS-related audits.
+  'is-on-https',
+  'redirects-http'
```
I don't think there is a way to run HTTPS-related tests on CI. Both tests pass on the preview and staging servers, so I just pretend they passed on CI.

petebacondarwin replied:
should we have a separate run for the staging site, which does test for https?

gkalpak replied:
You mean after we deploy to staging? Actually, if we are fine running this test after deployment (and not during the "test" phase), then we can run against the deployed versions (either preview or staging).

petebacondarwin replied:
I think it is useful to see in the preview if it has changed.
It is a bit too late to fail after it is deployed to staging, right?

gkalpak replied:
Never is too late to fail :smiley: (But it will remain deployed nevertheless, which is what we want anyway).

I made the changes so that `test-pwa-score` is run on CI for both the local build (possibly superfluous - up for debate) and the deployed version (if there is one - i.e. if the PR touched `aio` or if we are in upstream master). For both cases the min. expected score is 90, but we can tweak it per case, if necessary.

The only thing that is a little odd is that the build can fail even if the deployment to staging succeeded, but I don't think this is a major issue (and it is also not likely to happen if we keep the test against the local build).

mary-poppins commented:
The angular.io preview for 3c3a7b4 is available [here][1].

[1]: https://pr15688-3c3a7b45b2461de755c2e99ccf1f6d35e057e87a.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 6135028 is available [here][1].

[1]: https://pr15688-6135028092cd1306dfa3a1f81c6c6799daede01d.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 25e4912 is available [here][1].

[1]: https://pr15688-25e49120c12d50b02a36827f22a7009a804b9664.ngbuilds.io/

mary-poppins commented:
The angular.io preview for afd8388 is available [here][1].

[1]: https://pr15688-afd8388f41a45bce6926218ccdf7db65017e5fc3.ngbuilds.io/

gkalpak commented:
This is ready for review.
Caution! It touches files outside aio (althouth still aio-related :grin:).

petebacondarwin commented on aio/scripts/test-pwa-score.js:
```diff
@@ -0,0 +1,73 @@
+  ignoreAudits(IGNORED_AUDITS, config.aggregations);
+
+  launchChromeAndRunLighthouse(URL, FLAGS, config).
+    then(results => getScore(results)).
```
could be `then(getScore).`

gkalpak replied:
I know. I really struggled with this one :grin:
`getScore` is shorter, but `results => getScore(results)` fits in better with the following line 😞

petebacondarwin replied:
fair enough

gkalpak replied:
Changed it.

petebacondarwin commented on aio/scripts/test-pwa-score.js:
```diff
@@ -0,0 +1,73 @@
+  aggregations.forEach(aggregation =>
+    aggregation.items.forEach(item =>
+      ignoredAudits.map(key => item.audits[key]).forEach(audit =>
+        audit && (audit.expectedValue = !audit.expectedValue))));
```
It looks to me that rather than ignoring the audit value you are toggling its expectation?
What happens if it passes when we expected it to fail?

gkalpak replied:
Toggling the expectation keeps the results more close to what it would have been if it actually passed. That said, I am not super happy with this hack. I don't expect there to be frequent changes in this part of the code, so I am OK with the hack.

But it is probably even better to run against the deployed versions as discussed [here][1].

[1]: https://github.com/angular/angular/pull/15688#discussion_r109281555

gkalpak replied:
Now the tests are run against both the local build (without the HTTPS audits) and the deployed version (with all audits).
See https://github.com/angular/angular/pull/15688#discussion_r109284220.

petebacondarwin commented on aio/src/app/sw-updates/sw-update-notifications.service.spec.ts:
```diff
@@ -158,6 +172,12 @@ describe('SwUpdateNotificationsService', () => {
 });
 
 // Mocks
+class MockGlobal {
```
or `TestGlobal`? :-P

gkalpak replied:
I have a plan to overrule @wardbell and change everythign to `Mock` 😛

petebacondarwin replied:
He who writes the tests :-)

petebacondarwin commented on aio/src/styles/main.scss:
```diff
@@ -2,7 +2,7 @@
 @import './ng-io-theme';
 
-// import global variables / 
+// import global variables /
```
why is there a slash on the end of this line?

gkalpak replied:
No idea. Probably by accident.

gkalpak replied:
Removed.

petebacondarwin reviewed (commented):
A few minor comments.
An answer about the ignoreAudits comment would be good before approval.

gkalpak committed:
ci(aio): convert `deploy-staging` to a shell script

gkalpak commented:
@petebacondarwin , PTAL.

mary-poppins commented:
The angular.io preview for 8d271f0 is available [here][1].

[1]: https://pr15688-8d271f02678a85995ec0722671a59f9a89849ac2.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 3163740 is available [here][1].

[1]: https://pr15688-316374091aea53ff90b302f5e5645279fd479f59.ngbuilds.io/

IgorMinar commented:
@gkalpak please rebase

mary-poppins commented:
The angular.io preview for dbfc585 is available [here][1].

[1]: https://pr15688-dbfc585e0d8594caf97eb42c942d3597f7521aca.ngbuilds.io/

gkalpak commented:
@IgorMinar, @petebacondarwin: Rebased on master and added a fixup commit per our discussion: https://github.com/angular/angular/commit/14b649b66812261b3bc89d2962915ecbce0994a7
Waiting for Travis before squashing.

mary-poppins commented:
The angular.io preview for 14b649b is available [here][1].

[1]: https://pr15688-14b649b66812261b3bc89d2962915ecbce0994a7.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 57b6a14 is available [here][1].

[1]: https://pr15688-57b6a14cd34a0c0de690e4ec3d46f5dde0d3b193.ngbuilds.io/

petebacondarwin commented:
All is good for merging this. We just need @IgorMinar  to approve the change inside `root`.
